tiledb/sm/misc/type_casts.h

```diff
@@ -0,0 +1,146 @@
+/**
+ * @file type_casts.h
```

Comment:
This could be named safe_integral_casts.h to be more precise.

Reply:
My thought was to leave it slightly vague in case we added more type casting functions. Happy to make it more specific if you prefer.

Reply:
I think specific might be better here as type can be many many things.